libs/blockscout-auth/src/auth.rs

```diff
@@ -0,0 +1,347 @@
+use cookie::Cookie;
+use reqwest::{
+    header::{HeaderMap, HeaderValue},
+    StatusCode,
+};
+use serde::Deserialize;
+use std::collections::HashMap;
+use thiserror::Error;
+use tonic::{codegen::http::header::COOKIE, metadata::MetadataMap};
+use url::Url;
+
+const JWT_TOKEN_NAME: &str = "_explorer_key";
+const CSRF_TOKEN_NAME: &str = "x-csrf-token";
+const API_KEY_NAME: &str = "api_key";
+
+#[derive(Debug)]
+pub struct AuthSuccess {
+    pub user_id: String,
+}
+
+#[derive(Debug, Clone, Deserialize)]
+#[serde(untagged)]
+enum BlockscoutAuthResponse {
+    Success(BlockscoutAuthSuccessResponse),
+    Error(BlockscoutAuthErrorResponse),
+}
+
+#[derive(Debug, Clone, Deserialize)]
+struct BlockscoutAuthErrorResponse {
+    message: String,
+}
+
+#[allow(dead_code)]
+#[derive(Debug, Clone, Deserialize)]
+struct BlockscoutAuthSuccessResponse {
+    avatar: String,
+    email: String,
+    id: i64,
+    name: String,
+    nickname: String,
+    uid: String,
+    watchlist_id: i64,
+}
+
+impl From<BlockscoutAuthSuccessResponse> for AuthSuccess {
+    fn from(value: BlockscoutAuthSuccessResponse) -> Self {
+        Self {
+            user_id: value.id.to_string(),
+        }
+    }
+}
+
+#[derive(Error, Debug)]
+pub enum Error {
+    #[error("invalid jwt token: {0}")]
+    InvalidJwt(String),
+    #[error("invalid csrf token: {0}")]
+    InvalidCsrfToken(String),
+    #[error("user is unauthorized: {0}")]
+    Unauthorized(String),
+    #[error("blockscout invalid response: {0}")]
+    BlockscoutApi(String),
+    #[error("internal error: {0}")]
+    InternalError(String),
+}
+
+pub async fn auth_from_metadata(
+    metadata: &MetadataMap,
+    is_safe_http_method: bool,
+    blockscout_host: &Url,
+    blockscout_api_key: Option<&str>,
+) -> Result<AuthSuccess, Error> {
+    let jwt = extract_jwt(metadata)?;
+    let csrf_token = if is_safe_http_method {
+        None
+    } else {
+        Some(extract_csrf_token(metadata)?)
+    };
+    auth_from_tokens(&jwt, csrf_token, blockscout_host, blockscout_api_key).await
+}
+
+pub async fn auth_from_tokens(
+    jwt: &str,
+    csrf_token: Option<&str>,
+    blockscout_host: &Url,
+    blockscout_api_key: Option<&str>,
+) -> Result<AuthSuccess, Error> {
+    let headers = build_http_headers(jwt, csrf_token)?;
+    let client = reqwest::ClientBuilder::new()
+        .default_headers(headers)
+        .build()
+        .map_err(|e| Error::InternalError(e.to_string()))?;
+    let mut url = blockscout_host
+        .join("/api/account/v1/authenticate")
+        .expect("should be valid url");
+    url.set_query(
+        blockscout_api_key
+            .map(|api_key| format!("{API_KEY_NAME}={api_key}"))
+            .as_deref(),
+    );
+
+    let response = if csrf_token.is_some() {
+        client.post(url)
+    } else {
+        client.get(url)
+    }
+    .send()
+    .await
+    .map_err(|_| Error::BlockscoutApi("failed to connect".to_string()))?;
```

Comment:
Nit: may be a little clearer as avoid building a new client from builder which may potentially result in error
```suggestion
    let headers = build_http_headers(jwt, csrf_token)?;
    let client = reqwest::Client::new();
    let response = if csrf_token.is_some() {
        client.post(url)
    } else {
        client.get(url)
    }
    .headers(headers)
    .send()
    .await
    .map_err(|_| Error::BlockscoutApi("failed to connect".to_string()))?;
```